CesiumGltf/include/CesiumGltf/SharedAssetDepot.h

```diff
@@ -0,0 +1,363 @@
+
+      const AssetType* operator()(const AssetContainer<AssetType>* container) {
+        return &container->asset;
+      }
+    };
+
+    return std::visit(Operation{}, this->contents);
+  }
+
+  AssetType* operator->() { return this->get(); }
+  const AssetType* operator->() const { return this->get(); }
+  AssetType& operator*() { return *this->get(); }
+  const AssetType& operator*() const { return *this->get(); }
+
+  bool operator==(const SharedAsset<AssetType>& other) const {
+    return other.get() == this->get();
+  }
+
+  bool operator==(SharedAsset<AssetType>& other) {
+    return other.get() == this->get();
+  }
+
+  bool operator!=(const SharedAsset<AssetType>& other) const {
+    return other.get() != this->get();
+  }
+
+  bool operator!=(SharedAsset<AssetType>& other) {
+    return other.get() != this->get();
+  }
+
+  /**
+   * Copy assignment operator for SharedAsset.
+   */
+  void operator=(const SharedAsset<AssetType>& other) {
+    if (*this != other) {
+      // Decrement this reference
+      this->maybeChangeCounter(-1);
+      this->contents = other.contents;
+      // Increment the new reference
+      this->maybeChangeCounter(1);
+    }
+  }
+
+  /**
+   * Move assignment operator for SharedAsset.
+   */
+  void operator=(SharedAsset<AssetType>&& other) noexcept {
+    if (*this != other) {
+      this->maybeChangeCounter(-1);
+      this->contents = std::move(other.contents);
+      other.contents = nullptr;
+    }
+  }
+
+  ~SharedAsset() { this->maybeChangeCounter(-1); }
+
+  /**
+   * Copy constructor.
+   */
+  SharedAsset(const SharedAsset<AssetType>& other) {
+    contents = other.contents;
+    this->maybeChangeCounter(1);
+  }
+
+  /**
+   * Move constructor.
+   */
+  SharedAsset(SharedAsset<AssetType>&& other) noexcept {
+    contents = std::move(other.contents);
+    other.contents = nullptr;
+  }
+
+private:
+  SharedAsset(AssetContainer<AssetType>* container) : contents(container) {
+    this->maybeChangeCounter(1);
+  }
+
+  void maybeChangeCounter(int amt) {
+    struct Operation {
+      int amt;
+
+      void operator()(AssetContainer<AssetType>* asset) {
+        if (asset != nullptr) {
+          asset->counter += this->amt;
+          if (asset->counter <= 0) {
+            asset->parent->remove(asset->assetId);
+          }
+        }
+      }
+
+      void operator()([[maybe_unused]] AssetType& asset) {}
+    };
+
+    std::visit(Operation{amt}, this->contents);
+  }
+
+  // A SharedAssetRef might point to an asset the SharedAssetDepot, or it
+  // might not. If it doesn't, we own this asset now.
+  AssetContents contents;
+
+  friend class SharedAssetDepot;
+  friend class AssetContainer<AssetType>;
+  friend class SingleAssetDepot<AssetType>;
+};
+
+/**
+ * Contains one or more assets of the given type.
+ * @tparam AssetType The type of asset stored in this depot.
+ */
+template <typename AssetType> class SingleAssetDepot {
```

Comment:
It's the class that stores all assets of a single type, is what I'm trying to get across. The idea here is that we'll have multiple types of assets eventually in this depot (external metadata schemas being next on the agenda), and so the tileset will have one asset depot that can contain multiple types of assets. It's a little clunky but I was trying to avoid ending up with separate `getImagesSharedAssetDepot`, `getMetadataSchemasSharedAssetDepot`, `getBuffersSharedAssetDepot` etc methods on the TilesetContentManager.